c2rust-analyze/src/util.rs

```diff
@@ -182,6 +184,7 @@ fn builtin_callee<'tcx>(
 pub fn lty_project<'tcx, L: Debug>(
     lty: LabeledTy<'tcx, L>,
     proj: &PlaceElem<'tcx>,
+    adt_func: impl Fn(AdtDef<'tcx>, Field) -> LabeledTy<'tcx, L>,
```

Comment:
This closure should receive the `VariantIdx` too (assuming you decide to keep the `VariantIdx` as part of the key), so that the closure can essentially just be a lookup function and doesn't need to contain any interesting logic.

Reply:
`VariantIdx` is no longer part of the key

Reply:
Is this because `ProjectionElem::Field` can't be called on an `enum`, only a `struct`?

Reply:
`ProjectionElem::Field` can be called after using `ProjectionElem::Downcast` to convert an enum to a specific variant of that enum.  For example, the MIR for this code:
```Rust
fn f(x: Option<i32>) -> i32 {
    match x {
        Some(y) => y,
        None => 0,
    }
}
```
includes this line in the `Some` case:
```Rust
_3 = ((_1 as Some).0: i32);
```
The `Place` on the RHS consists of the local `_1`, a `ProjectionElem::Downcast` (`... as Some`), and a `ProjectionElem::Field` (`(...).0`).

Reply:
In that above case, what would the `f` be in the `ProjectionElem::Field(f, _)`?

Reply:
0, matching the `.0` in the pretty-printed version.

This is more interesting with `Result` the two cases you get are these:
```Rust
// Ok case
_3 = ((_1 as Ok).0: i32);
// Err case
_4 = ((_1 as Err).0: u8);
```
Each variant has only one field, so both `Field`s use index 0.  They're distinguished instead by the `VariantIdx` used in the `ProjectionElem::Downcast`.